### PR TITLE
Null safety migration

### PR DESCRIPTION
**Pull request by DevNico**
General question since ValueNotifier has a non null generic i have implemented a second `useState` and `useValueNotifier` hook for nullable types: `useNullableState` and `useNullableValueNotifier`. Do you think this is the way to go or do you have a better idea? @rrousselGit

I also just removed all tests that are no longer necessary because of null safety.

### CODE REVIEW

**Comment by DevNico** [comment]
The tests seem to be failing due to find.text working different in CI? Locally all tests pass.

**Comment by rrousselGit** [comment]
I don't quite understand. What is the purpose of `useNullableState`? Why not do `useState<Model?>(null)`

**Comment by DevNico** [comment]
It would be to make the initialState optional but I agree adding another hook would be unnecessary but the initialState wouldn't be optional anymore.

**Comment by rrousselGit** [comment]
Indeed. I think removing them is better

**Comment by rrousselGit** [comment]
Could you fix the CI too? Thanks!

**Comment by DevNico** [comment]
I feel like the coverage diff is acceptable. LGTM @rrousselGit?

**Comment by rrousselGit** [comment]
@DevNico could you give me the rights to make edits on your PR? Thanks!

File: lib/src/animation.dart [hunk 1]
```diff
@@ -70,16 +70,16 @@ class _AnimationControllerHook extends Hook<AnimationController> {
...
-    this.upperBound,
-    this.vsync,
-    this.animationBehavior,
-    List<Object> keys,
+    required this.initialValue,
+    required this.lowerBound,
+    required this.upperBound,
+    required this.vsync,
+    required this.animationBehavior,
+    List<Object>? keys,
```
Comment (rrousselGit) [inline lib/src/animation.dart:78]: Keys needs to be `List<Object?>? keys`

File: lib/src/async.dart [hunk 2]
```diff
@@ -9,7 +9,7 @@ part of 'hooks.dart';
 ///   * [Future], the listened object.
 ///   * [useStream], similar to [useFuture] but for [Stream].
 AsyncSnapshot<T> useFuture<T>(Future<T> future,
-    {T initialData, bool preserveState = true}) {
+    {T? initialData, bool preserveState = true}) {
```
Comment (rrousselGit) [inline lib/src/async.dart:12]: initialData probably should be `required T initialData`

It's a breaking change, but matches with how `FutureProvider`/`StreamProvider` migrated

File: lib/src/async.dart [hunk 3]
```diff
@@ -19,7 +19,7 @@ class _FutureHook<T> extends Hook<AsyncSnapshot<T>> {
 
   final Future<T> future;
```
Comment (rrousselGit) [inline lib/src/async.dart:20]: The Future (and Stream with useStream) should be nullable

File: lib/src/async.dart [hunk 4]
```diff
@@ -29,14 +29,18 @@ class _FutureStateHook<T> extends HookState<AsyncSnapshot<T>, _FutureHook<T>> {
   /// An object that identifies the currently active callbacks. Used to avoid
   /// calling setState from stale callbacks, e.g. after disposal of this state,
   /// or after widget reconfiguration to a new Future.
-  Object _activeCallbackIdentity;
-  AsyncSnapshot<T> _snapshot;
+  Object? _activeCallbackIdentity;
+  late AsyncSnapshot<T> _snapshot;
```
Comment (rrousselGit) [inline lib/src/async.dart:33]: Prefer nullable properties instead of `late`.

That keyword is dangerous

File: lib/src/async.dart [hunk 5]
```diff
@@ -158,38 +165,35 @@ class _StreamHookState<T> extends HookState<AsyncSnapshot<T>, _StreamHook<T>> {
   }
 
   void _subscribe() {
-    if (hook.stream != null) {
```
Comment (rrousselGit) [inline lib/src/async.dart:161]: those null checks shouldn't be removed

File: lib/src/async.dart [hunk 6]
```diff
@@ -158,38 +165,35 @@ class _StreamHookState<T> extends HookState<AsyncSnapshot<T>, _StreamHook<T>> {
...
   }
 
   @override
   AsyncSnapshot<T> build(BuildContext context) {
     return _summary;
   }
 
-  AsyncSnapshot<T> initial() =>
-      AsyncSnapshot<T>.withData(ConnectionState.none, hook.initialData);
+  AsyncSnapshot<T> initial() => hook.initialData != null
```
Comment (rrousselGit) [inline lib/src/async.dart:194]: to remove after initialData is made required

File: lib/src/async.dart (hunk 2 above)
Comment (DevNico) [reply lib/src/async.dart:12]: would you like to keep it as a named parameter or what do you think about either both `future` and `T initialData` as positioned parameter or both as named?

File: lib/src/async.dart (hunk 3 above)
Comment (DevNico) [reply lib/src/async.dart:20]: > @DevNico could you give me the rights to make edits on your PR? Thanks!

The checkbox to allow edits from maintainers is already enabled or do you mean something else?

**Comment by DevNico** [comment]
> @DevNico could you give me the rights to make edits on your PR? Thanks!

The checkbox to allow edits from maintainers is already enabled or do you mean something else?

File: lib/src/async.dart (hunk 4 above)
Comment (DevNico) [reply lib/src/async.dart:33]: So would you want useFuture to return a nullable AsyncSnaphot (`AsyncSnapshot<T>? useFuture<T>`) or should the build method return a non nullable snapshot since it should be initialized anyway?

Comment (DevNico) [reply lib/src/async.dart]: How would you feel about
`late AsyncSnapshot<T> _summary = AsyncSnapshot<T>.withData(ConnectionState.none, hook.initialData);`?

Comment (rrousselGit) [reply lib/src/async.dart]: The return value must be non-nullable



> How would you feel about
> `late AsyncSnapshot<T> _summary = AsyncSnapshot<T>.withData(ConnectionState.none, hook.initialData);`?

That's fine, it's mostly uninitialized late variables that are scary

Then we can do the same thing with other hooks, like useStreamController

**Comment by rrousselGit** [comment]
Weird, I've tried pushing some changes to your branch but the push was rejected

**Comment by rrousselGit** [comment]
Thanks! I'll take over from here and will try to merge & publish it later today.

Do you happen to have a twitter handle to share? I'd want to thank you for your help 😄

**Comment by DevNico** [comment]
Sure my handle is @DevNicoS even though I am not very active on Twitter 😅

**Comment by rrousselGit** [comment]
Alright I'm going to release this. Thanks again for the help!
